### ISSUE TITLE
How to render markers on map fetched from remote backend API?

### ISSUE THREAD

**Issue by sayedatif** [react-community/react-native-maps#384]
I am trying to render markers from remote API. I tried fetching data using fetch and tried logging it. I can see it in console log. so how can i use it to render markers on map? remote API is fully functional

fetchData: function(){
    fetch("http://www.example.com/endpoint", {
    method: "POST",
    body: JSON.stringify({
      tag:"get_service_all",
      lat:"19.34",
      lon:"72.845",
    })
  })
  .then((response) => response.json())
  .then((responseData) => console.log(responseData))
  .done();
},

and tried mounting it using

componentDidMount: function() {
    this.fetchData();
}

so how can i setState? does anybody have previous experience on this? Any help is appreciated and thanks alot in advance.

**Comment by chrisv95** [comment]
I have some experience.
You should put your map in the render() function; and it should look somewhat like this:

    <MapView ....>
    {this.state.markers.map(marker => {
	return (
		<MapView.Marker
		  coordinate={marker.latlng}
		/>
	);
    })}
    </MapView>`

then, your getInitialState must be something like

    getInitialState() {
		return {
			markers: []
		};
    },

and then, once you need to fetch your data (I fetched the data in componentDidMount), after you fetched and have the elements available, you do this:

    var elements = elements.map(function(e) {
    	return {
    		id: e[0],
    		latlng: {
    			latitude: e[2],
    			longitude: e[3]
    		}
    	}
    });
    this.setState({
    	markers: elements
    });

then, because the state of 'markers' has changed, the render() function will re-render the map, including the markers.
For this to work, make sure that your source file is a parsable file (preferably JSON, or you'll need to change some things for this to work) and that it is an array of contents (in my example, it's an array of arrays given my data source specification, but it could also be an array of objects -- you'd just have to change your code a bit)

**Comment by sayedatif** [comment]
can you help me with some sample code ?

**Comment by chrisv95** [comment]
I think react-native-maps's documentation + my sample code above will help you ;-)

**Comment by sayedatif** [comment]
hey @chrisv95 i tried implementing your code. here is the file

var React = require('react');
var ReactNative = require('react-native');

var {
  Alert,
  StyleSheet,
  Dimensions,
} = ReactNative;

var MapView = require('react-native-maps');
const windowsWidth = Dimensions.get('window').width;
const windowsHeight = Dimensions.get('window').height;


var RanMap = React.createClass({
  getInitialState() {
    return {
        markers: []
    };
  },

  componentDidMount: function(){
    fetch("http://www.example.com/endpoints", {
    method: "POST",
    body: JSON.stringify({
      tag:"get_service_all",
      lat:"19.34",
      lon:"72.845",
    })
  })
  .then((response) => response.json())
  .then((responseData) => {
    return responseData
  })
  .done();
  },

  render() {
    <MapView
      ref="map"
      style={styles.map}
      >
    {this.state.markers.map(marker => {
        return (
          <MapView.Marker
          coordinate={marker.latlng}
          />
        );
      })
   }
    </MapView>
  }
})

var elements = elements.map(function(e) {
    return {
        id: e[0],
        latlng: {
            latitude: e[2],
            longitude: e[3]
        }
    }
});
this.setState({
    markers: elements
});

var styles = StyleSheet.create({
  map: {
    position: 'absolute',
    height: windowsHeight/2 + windowsHeight/3,
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
  }
});

module.exports = RanMap;

but ran into error cannot read property map of undefined. I also think code need some rectification can you please rectify it for me? thanks

**Comment by chrisv95** [comment]
Your code

    var elements = elements.map(function(e) {
    return {
    id: e[0],
    latlng: {
    latitude: e[2],
    longitude: e[3]
    }
    }
    });
    this.setState({
    markers: elements
    });

should be added to the Promise that is returned on fetch(): then(responseData);
then, connect the latitudes & longitudes in your responseData to the mapping by changing `elements.map` into `responseData.map`

**Comment by sayedatif** [comment]
hey @chrisv95  can you provide me with some tutorial? because i am not getting things right  as i am new to react-native .

**Comment by sayedatif** [comment]
hey @chrisv95 
here is my backend API JSON file
`{
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "hits": {
    "total": 1,
    "max_score": 1,
    "hits": [
      {
        "_index": "service",
        "_type": "service",
        "_id": "ac8c5edd-1aad-406f-b476-012c6e940c1a",
        "_score": 1,
        "_source": {
          "service_owner_id": "2",
          "service_name": "-----",
          "service_email": "mk@gmail.com",
          "service_contact_name": "-----",
          "service_landline_number": "12345",
          "service_mobile_number": "1234567890",
          "service_address": "vdvml",
          "service_listingType": 1,
          "service_avg_rating": 3.5,
          "sarvice_ratingcount": 10,
          "service_serviceType": 1,
          "service_working_hour": [
            ""
          ],
          "service_subcat_ids": "2,3",
          "service_description": "hi I am ------ ",
          "service_id": "ac8c5edd-1aad-406f-b476-012c6e940c1a",
          "service_logo_url": "",
          "service_gallery_url": "",
          "service_location": {
            "lat": 19.34,
            "lon": 72.845
          },
          "service_subcategory": "CAT GER ",
          "service_review": []
        }
      }
    ]
  },
  "took": 1,
  "timed_out": false
}`

and i want to fetch the marker from it how can i do it?

**Comment by sayedatif** [comment]
hey @chrisv95 can you just tell me why i am error:facing no position in marker options

`fetchData: function(){
    fetch("http://www.example.com/endpoints, {
    method: "POST",
    body: JSON.stringify({
      tag:"get_service_all",
      lat:"19.34",
      lon:"72.845",
    })
  })
  .then((response) => response.json())
  .then((responseData) => {
    this.setState({
      markers: responseData.hits.hits.map(function(hit){
      const service_location = hit._source.service_location;
      return {
        latlng{
          latitude: service_location.lat,
          longitude: service_location.lon
          }
        }
      })
    })
    console.log(this.state.markers)
  })`

here is the render()

`{this.state.markers.map(marker => {
            return(
              <MapView.Marker coordinate={marker.latlng} />
            );
          })}`

![mapserror](https://cloud.githubusercontent.com/assets/6937774/16841437/d6b0102c-49f6-11e6-838b-eacb66ddb474.PNG)
can you help on this?

**Comment by sayedatif** [comment]
@KGALLET hey i tried putting ID there but no success the warning still appears in the console

here is the code.

```
.then((responseData) => {
      this.setState({
        markers: responseData.hits.hits.map(function(hit){
          const service_location = hit._source.service_location;
          return {
            id: hit._source.service_id,
            latlng:{
              latitude: service_location.lat,
              longitude: service_location.lon
            }
          }
        })
      })
```
can you help me out how to get rid of this warning?

**Comment by KGALLET** [comment]
@sayedatif ok so you have the id in the response of your request.
Now in the render you have to do something like (according to what you had already done) : 

```
{this.state.markers.map(marker => {
            return(
              <MapView.Marker coordinate={marker.latlng} key={marker.id} />
            );
          }
     )
}
```

**Comment by sayedatif** [comment]
hey kevin can you help me on this http://stackoverflow.com/questions/38496077/how-can-i-get-markers-within-specific-radius-near-user-current-location-in-react

**Comment by KGALLET** [comment]
Im gonna check 👍

**Comment by sayedatif** [comment]
hey kevin my markers are mounting on app start instead i want only those markers which are near user location or they come inside the radius specified in <MapView.Circle>. I am using ComponentDidMount() which is called after getInitialState() what can i do instead of it?

**Comment by KGALLET** [comment]
In your endpoint, do you have any option to set the geoLocation and a radius?
This is how I did. 
```
const MyListComponentQuery = container({
  variables: (props) => {
    return {
      after: props.after,
      text: props.text,
      categories: props.categories,
      geoLocation: props.geoLocation,
      radius: props.radius,
      datesRange: props.datesRange,
      dates: props.dates
    }
  },
  query: `
    query MyQuery ($after: String, $text: String, $categories: [String],
        $geoLocation: String, $radius: Int, $datesRange: [DateTime], $dates: [Date]) {
      events: culturalEvents(
         first: 30
         after: $after
         text: $text
         categories: $categories
         geoLocation: $geoLocation
         radius: $radius
         datesRange: $datesRange
         dates: $dates
         sortOn: "start_date"
      ) {
        pageInfo {
          endCursor
          hasNextPage
        }
        edges {
          node {
            ${MyItemComponent.getFragment('event')}
          }
        }
      }
    }
  `,
})(MyListComponent)
```

I have a prop for radius and one for the geoLocation, that's why I can use the MapView.Circle

**Comment by sayedatif** [comment]
yes i have 
here is my code.
```
fetchData: function(){
    fetch("http://www.example.com/endpoints", {
      method: "POST",
      body: JSON.stringify({
        tag:"get_service_all",
        lat:"19.34",
        lon:"72.845",
        distance: "3000",
      })
    })
```
distance works as a radius

**Comment by KGALLET** [comment]
Ok so you are supposed to set lat, lon with the user location and distance with the radius you want too, in order to only get "markers" near the user position. :)

**Comment by sayedatif** [comment]
so how can i do it? i want only those marker which are near user but my markers are getting rendered even if they are not in the specified radius you can see the image
![img-20160721-wa0002_1469091349918](https://cloud.githubusercontent.com/assets/6937774/17017006/3957c780-4f4f-11e6-8f99-4004584f8b9f.jpg)

**Comment by KGALLET** [comment]
I think the problem is your request... I can't help you more since I don't have access to it :/

**Comment by sayedatif** [comment]
hey kevin can i comment my api here so that you get some idea?

**Comment by sayedatif** [comment]
hey @KGALLET can you help me on this http://stackoverflow.com/questions/38524371/i-am-trying-to-get-markers-near-user-location-by-calculating-distance-using-have

**Comment by KGALLET** [comment]
It seems someone answered u :)

**Comment by sayedatif** [comment]
hey @KGALLET  can you tell me what is the problem in  this code

```
getWithinDistance: function(){
    var selectedMarker = [];
    var lat1 = this.state.region.latitude;
    var lon1 = this.state.region.longitude;
    for(var i = 0 ; i <= 3 ; i++)
      {
      var lat2 = this.state.a[i].latitude;
      var lon2 = this.state.a[i].longitude;
      var R = 6371;
      var dLat = (lat2-lat1) * (Math.PI/180);
      var dLon = (lon2-lon1) * (Math.PI/180);
      var a = Math.sin(dLat/2) * Math.sin(dLat/2) + Math.cos((lat1) * (Math.PI/180)) * Math.cos((lat2) * (Math.PI/180)) * Math.sin(dLon/2) * Math.sin(dLon/2);
      var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
      var d = R * c;
      if(d <= 2.00){
        var latlng = {
          latitude: lat2,
          longitude: lon2
        }
        selectedMarker.push(latlng);
      }
      }

      var at = selectedMarker.map(marker => {
        return(
          <MapView.Marker coordinate={marker.latlng} />
        );
      })
      
  },
```

it is showing no position in marker options what might be the issue i tried logging but i couldn't find out what is wrong

**Comment by KGALLET** [comment]
Your code is not easy to understand when you don't know what is the goal.
I'm gonna check next week, not the time for the moment

**Comment by alvelig** [comment]
Not related

**Comment by vipin-developer** [comment]
I am not able to render markers on map fetched from remote backend API but only after creating apk.
The app is getting crashed, earlier it was working smoothly but suddenly it stopped.
the error is like "uncaught error: require unknown module 'undefined ' ".
Although it is working fine in emulators.
this strange issue I am facing from last 2 weeks.
react-native-maps: "0.22.0"
"react": "16.5.0",
    "react-native": "https://github.com/expo/react-native/archive/sdk-32.0.0.tar.gz",

**Comment by vipin-developer** [comment]
And the expo version is 
"expo": "32.0.0",
